medmodels/data_synthesis/synthesizer.py

```diff
@@ -0,0 +1,124 @@
+"""Abstract class for synthesizers."""
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Dict, Optional, Type
+
+import torch
+from torch import nn
+
+from medmodels.data_synthesis.synthesizer_model import SynthesizerModel
+from medmodels.medrecord.medrecord import MedRecord
+
+if TYPE_CHECKING:
+    from medmodels.data_synthesis.builder import SynthesizerBuilder
+
+
+class Synthesizer(nn.Module, metaclass=ABCMeta):
+    """Synthesizer is an abstract class that serves as a blueprint for synthesizers.
+
+    It ensures the correct instantiation, training, and persistence of models.
+    """
+
+    _preprocessor: Type[nn.Module]
+    _postprocessor: Type[nn.Module]
+    device: torch.device
+
+    def __init__(
+        self,
+        preprocessor: Type[nn.Module],
+        postprocessor: Type[nn.Module],
+    ) -> None:
+        """Initializes the synthesizer with the required components.
+
+        Args:
+            preprocessor (Type[nn.Module]): Type of preprocessor module.
+            postprocessor (Type[nn.Module]): Type of postprocessor module.
+        """
+        super(Synthesizer, self).__init__()
+        self._preprocessor = preprocessor
+        self._postprocessor = postprocessor
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```

Comment:
the device should be selectable, not automatically use the gpu if the user might not want to